pkg/prometheus/agent/operator.go

```diff
@@ -564,6 +581,82 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 
 	p := pobj.(*monitoringv1alpha1.PrometheusAgent)
 	p = p.DeepCopy()
+	if p.Spec.Mode != nil && *p.Spec.Mode == "DaemonSet" {
+		err = c.syncDaemonSet(ctx, key, p)
+		c.reconciliations.SetStatus(key, err)
+		return err
+	}
+	err = c.syncStatefulSet(ctx, key, p)
+	c.reconciliations.SetStatus(key, err)
+	return err
+}
+
+func (c *Operator) syncDaemonSet(ctx context.Context, key string, p *monitoringv1alpha1.PrometheusAgent) error {
+	if err := k8sutil.AddTypeInformationToObject(p); err != nil {
+		return fmt.Errorf("failed to set Prometheus type information: %w", err)
+	}
+
+	logger := log.With(c.logger, "key", key)
+
+	if p.Spec.Paused {
+		level.Info(logger).Log("msg", "the resource is paused, not reconciling")
+		return nil
+	}
+
+	level.Info(logger).Log("msg", "sync prometheus")
+
+	cg, err := prompkg.NewConfigGenerator(c.logger, p, c.endpointSliceSupported)
+	if err != nil {
+		return err
+	}
+
+	assetStore := assets.NewStoreBuilder(c.kclient.CoreV1(), c.kclient.CoreV1())
+	if err := c.createOrUpdateConfigurationSecret(ctx, p, cg, assetStore); err != nil {
+		return fmt.Errorf("creating config failed: %w", err)
+	}
+
+	tlsAssets, err := operator.ReconcileShardedSecret(ctx, assetStore.TLSAssets(), c.kclient, prompkg.NewTLSAssetSecret(p, c.config))
+	if err != nil {
+		return fmt.Errorf("failed to reconcile the TLS secrets: %w", err)
+	}
+
+	if err := c.createOrUpdateWebConfigSecret(ctx, p); err != nil {
+		return fmt.Errorf("synchronizing web config secret failed: %w", err)
+	}
+
+	dsetClient := c.kclient.AppsV1().DaemonSets(p.Namespace)
+
+	level.Debug(logger).Log("msg", "reconciling daemonset")
+
+	_, err = c.dsetInfs.Get(prompkg.KeyToDaemonSetKey(p, key))
+	exists := !apierrors.IsNotFound(err)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return fmt.Errorf("retrieving daemonset failed: %w", err)
+	}
+
+	dset, err := makeDaemonSet(
+		"daemonset",
```

Comment:
```suggestion
		p.Name,
```

Could you explain why we're using "daemonset" as the name here?

Reply:
For the statefulset we needed extra logix to handle shard names, but looks like we don't need that for daemonsets. Would it make sense to just remove this argument from `makeDaemonset`?